File: 2024/day01/javascript/solution.js
function getTuple(row) {
  const matches = row.match(/\d+/g);
  return [Number(matches[0]), Number(matches.slice(-1)[0])];
}

function part1(input) {
  rows = input.trim().split("\n").map(getTuple);
  left = rows.map((v) => v[0]).toSorted((a, b) => a - b);
  right = rows.map((v) => v[1]).toSorted((a, b) => a - b);
  zipped = left.map((l, i) => Math.abs(l - right[i]));
  return zipped.reduce((a, v) => a + v);
}

function countNums(input) {
  const map = new Map();

  for (const digit of input) {
    map.set(digit, (map.get(digit) ?? 0) + 1);
  }
  return map;
}

function part2(input) {
  rows = input.trim().split("\n").map(getTuple);
  left = rows.map((v) => v[0]);
  right = rows.map((v) => v[1]);
  rightCounts = countNums(right);
  values = left.map((num) =>
    rightCounts.has(num) ? num * rightCounts.get(num) : 0,
  );
  return values.reduce((a, v) => a + v);
}

module.exports = { part1, part2 };

File: 2024/day01/python/solution.py
import re
from collections import Counter


def get_part1_tuple(row):
    pattern = r"\d+"
    digit_strings = re.findall(pattern, row)
    return (int(digit_strings[0]), int(digit_strings[1]))


def part1(input_data):

    rows = [get_part1_tuple(row) for row in input_data.strip().split("\n")]
    left = sorted([values[0] for values in rows])
    right = sorted([values[1] for values in rows])
    zipped = [(lValue, rValue) for lValue, rValue in zip(left, right)]
    diffs = [abs(lValue - rValue) for lValue, rValue in zipped]
    return sum(diffs)


def part2(input_data):

    rows = [get_part1_tuple(row) for row in input_data.strip().split("\n")]
    left = [values[0] for values in rows]
    right = Counter([values[1] for values in rows])
    values = [num * right[num] for num in left]
    print(values)
    return sum(values)
